R/search-pv.R

```diff
@@ -96,31 +95,35 @@ request_apply <- function(ex_res, method, query, base_url, arg_list, api_key, ..
   if (req_pages < 1) {
     stop2("No records matched your query...Can't download multiple pages")
   }
-  if (matched_records > 10000) {
-    stop2(
-      "The API only allows you to download 10,000 records in a single query. ",
-      "Your query returned ", matched_records, " records. See <LINK> for ",
-      "how to get around this limitation."
-    )
-  }
-  if (req_pages > 10) {
-    stop2(
-      "The API only allows you to download 10 pages in a single query. ",
-      "Your query returned ", req_pages, " pages. See <LINK> for ",
-      "how to get around this limitation."
-    )
-  }
 
   tmp <- lapply(seq_len(req_pages), function(i) {
-    arg_list$opts$offset <- (i - 1) * arg_list$opts$size
     x <- one_request(method, query, base_url, arg_list, api_key, ...)
     x <- process_resp(x)
+
+    # now to page we need set the "after" attribute to where we left off
```

Comment:
This seems like such a weird pattern, where the paging requires us to get the last ID that is sorted on in a given iteration. Any idea if this is a permanent thing? I'd be curious to know if the ElasticSearch DB is supporting this kind of paging natively or if it's hacked together on patentsviews' end. Any idea? I couldn't find the source code for the latest API code easily.

Reply:
I agree it's a weird pattern!  I also looked for the source but couldn't find it. 

They did update the query language page mentioning the after parameter and the api throws errors if "offset" is passed so I'm guessing it's permanent.   ElasticSearch is new to me  but it looks like after is [supported natively](https://www.elastic.co/guide/en/elasticsearch/reference/current/paginate-search-results.html).  I'm wondering if it somehow let them remove the 10,000 row limit?   I'll have to see if ElasticSearch handles more than a primary sort field and what it uses as an after parameter.  Maybe even write a test case for it.

Reply:
Paging does seem to work with a secondary sort, passing an array of both keys in the "after" parameter.

curl -X 'POST' \
  'https://search.patentsview.org/api/v1/patent/' \
  -H 'accept: application/json' \
  -H 'X-Api-Key: my-api-key' \
  -H 'Content-Type: application/json' \
  -d '{
  "f": ["patent_id", "patent_date"],
  "o": {"size": 100, "after":["2007-01-02", "RE39455"]},
  "q": {"_gte":{"patent_date":"2007-01-02"}},
  "s": [{"patent_date": "asc"}, {"patent_id": "asc"}]
}'

Reply:
That same ElasticSearch page says 10,000 rows is the paging limit when using the native size and from attributes.  It looks like the after parameter gets around that which must be why the API team made the change.  Guessing the API change will be permanent.